[Merged by Bors] - feat: Cardinal of GLn with coefficients in a finite field

ThomasLanard opened the pull request:
---
We add the cardinal of the general linear group GLn with coefficients in a finite field.
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+    Fintype ((Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽)))ᶜ : Set (Fin n → 𝔽)) :=
+  Fintype.ofFinite _
+
+lemma complement_card {k : ℕ} (s : { s : Fin k → Fin n → 𝔽 // LinearIndependent 𝔽 s }):
+    Fintype.card ((Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽)))ᶜ : Set (Fin n → 𝔽)) =
+      (q) ^ n - (q) ^ k := by
+  rw [Fintype.card_compl_set, Fintype.card_fun, Fintype.card_fin]
+  simp only [SetLike.coe_sort_coe]
+  rw [card_eq_pow_finrank (K := 𝔽) (V := Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽))),
+    finrank_span_eq_card s.property, Fintype.card_fin]
```
```suggestion
lemma complement_card {k : ℕ} (s : { s : Fin k → Fin n → 𝔽 // LinearIndependent 𝔽 s }):
    Fintype.card ((Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽)))ᶜ : Set (Fin n → 𝔽)) =
      (q) ^ n - (q) ^ k := by
  simp only [Fintype.card_compl_set, Fintype.card_pi, Finset.prod_const, Finset.card_univ,
    Fintype.card_fin, SetLike.coe_sort_coe,
    card_eq_pow_finrank (K := 𝔽) (V := Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽))),
    finrank_span_eq_card s.property]
```

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+  rw [card_eq_pow_finrank (K := 𝔽) (V := Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽))),
+    finrank_span_eq_card s.property, Fintype.card_fin]
+
+def inductiveStepEquiv {k : ℕ} :
+    { s : Fin (k + 1) → Fin n → 𝔽 // LinearIndependent 𝔽 s } ≃
+      Σ s : { s : Fin k → Fin n → 𝔽 // LinearIndependent 𝔽 s },
+        ((Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽)))ᶜ : Set (Fin n → 𝔽)) where
+  toFun s := by
+    have := linearIndependent_fin_succ.mp s.property
+    use ⟨Fin.tail s.val, this.left⟩
+    exact ⟨s.val 0, this.right⟩
+  invFun s := by
+    use Fin.cons s.2.val s.1.val
+    exact linearIndependent_fin_cons.mpr ⟨s.1.property, s.2.property⟩
+  left_inv _ := by simp only [Fin.cons_self_tail, Subtype.coe_eta]
+  right_inv := fun ⟨_, _⟩ => by simp only [Fin.cons_zero, Subtype.coe_eta, Sigma.mk.inj_iff,
+    Fin.tail_cons, heq_eq_eq, and_self]
```
```suggestion
def inductiveStepEquiv {k : ℕ} :
    { s : Fin (k + 1) → Fin n → 𝔽 // LinearIndependent 𝔽 s } ≃
      Σ s : { s : Fin k → Fin n → 𝔽 // LinearIndependent 𝔽 s },
        ((Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽)))ᶜ : Set (Fin n → 𝔽)) where
  toFun s := by
    refine ⟨⟨Fin.tail s.val, (linearIndependent_fin_succ.mp s.property).left⟩,
      ⟨s.val 0, (linearIndependent_fin_succ.mp s.property).right⟩⟩
  invFun s := by
    refine ⟨Fin.cons s.2.val s.1.val, linearIndependent_fin_cons.mpr ⟨s.1.property, s.2.property⟩⟩
  left_inv _ := by simp only [Fin.cons_self_tail, Subtype.coe_eta]
  right_inv := fun ⟨_, _⟩ => by simp only [Fin.cons_zero, Subtype.coe_eta, Sigma.mk.inj_iff,
    Fin.tail_cons, heq_eq_eq, and_self]
```

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+  simp only [SetLike.coe_sort_coe]
+  rw [card_eq_pow_finrank (K := 𝔽) (V := Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽))),
+    finrank_span_eq_card s.property, Fintype.card_fin]
+
```
Since this is a `def` it will need a docstring. So add something like /-- Define equivalence from... -/ in the line above.

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+  right_inv := fun ⟨_, _⟩ => by simp only [Fin.cons_zero, Subtype.coe_eta, Sigma.mk.inj_iff,
+    Fin.tail_cons, heq_eq_eq, and_self]
+
+lemma card_LinearInependent_subtype {k : ℕ} (hk : k ≤ n) :
+    Fintype.card { s : Fin k → (Fin n → 𝔽) // LinearIndependent 𝔽 s } =
+      ∏ i : Fin k, ((q) ^ n - (q) ^ i.val) := by
+  induction' k with k ih
+  · simp only [Nat.zero_eq, LinearIndependent, Finsupp.total_fin_zero, LinearMap.ker_zero,
+    Fintype.card_ofSubsingleton, Finset.univ_eq_empty, Finset.prod_empty]
+  · simp only [Fintype.card_congr (inductiveStepEquiv n), Fintype.card_sigma, complement_card n, Finset.sum_const, Finset.card_univ, smul_eq_mul, ih (Nat.le_of_succ_le hk), mul_comm,
+    Fin.prod_univ_succAbove _ k, Fin.natCast_eq_last, Fin.val_last, Fin.succAbove_last,
+    Fin.coe_castSucc]
```
```suggestion
lemma card_LinearInependent_subtype {k : ℕ} (hk : k ≤ n) :
    Fintype.card { s : Fin k → (Fin n → 𝔽) // LinearIndependent 𝔽 s } =
      ∏ i : Fin k, ((q) ^ n - (q) ^ i.val) := by
  induction' k with k ih
  · simp only [Nat.zero_eq, LinearIndependent, Finsupp.total_fin_zero, LinearMap.ker_zero,
    Fintype.card_ofSubsingleton, Finset.univ_eq_empty, Finset.prod_empty]
  · simp only [Fintype.card_congr (inductiveStepEquiv n), Fintype.card_sigma, complement_card n,
    Finset.sum_const, Finset.card_univ, ih (Nat.le_of_succ_le hk), smul_eq_mul, mul_comm,
    Fin.prod_univ_succAbove _ k, Fin.natCast_eq_last, Fin.val_last, Fin.succAbove_last,
    Fin.coe_castSucc]
```

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+  left_inv _ := by simp only [Fin.cons_self_tail, Subtype.coe_eta]
+  right_inv := fun ⟨_, _⟩ => by simp only [Fin.cons_zero, Subtype.coe_eta, Sigma.mk.inj_iff,
+    Fin.tail_cons, heq_eq_eq, and_self]
+
```
The bot is complaining about the lines being too long. There is a 100 character limit. If you put `#lint` at the end of the file it will tell you any things that need fixing, although my suggestion should fix it.

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+    Fin.prod_univ_succAbove _ k, Fin.natCast_eq_last, Fin.val_last, Fin.succAbove_last,
+    Fin.coe_castSucc]
+
+lemma eq_matrix_basis (M : Matrix (Fin n) (Fin n) 𝔽) :
+    M = Basis.toMatrix (Pi.basisFun 𝔽 (Fin n)) (transpose M) := rfl
```
You can remove this, if you change the lemma below to what I've suggested.

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+
+/-- Equivalence between `GL n F` and `n` vectors of length `n` that are linearly independent. Given
+by sending a matrix to its coloumns. -/
+noncomputable def equiv_GL_linearindependent {F : Type*} [Field F] (hn : 0 < n) :
+    GL (Fin n) F ≃ { s : Fin n → (Fin n → F) // LinearIndependent F s } where
+  toFun M := ⟨transpose M, by
+    apply linearIndependent_iff_card_eq_finrank_span.2
+    rw [Set.finrank, ← rank_eq_finrank_span_cols, rank_unit]⟩
+  invFun M := by
+    apply GeneralLinearGroup.mk'' (transpose (M.1))
+    rw [eq_matrix_basis n (transpose (M.1)), transpose_transpose]
+    have : Nonempty (Fin n) := Fin.pos_iff_nonempty.1 hn
+    have hdim : Fintype.card (Fin n) = FiniteDimensional.finrank F (Fin n → F) := by
+      simp only [Fintype.card_fin, FiniteDimensional.finrank_fintype_fun_eq_card]
+    let b := basisOfLinearIndependentOfCardEqFinrank M.2 hdim
+    rw [show M = ⇑b by simp only [b, coe_basisOfLinearIndependentOfCardEqFinrank]]
+    have : Invertible ((Pi.basisFun F (Fin n)).toMatrix ⇑b) :=
+      (Pi.basisFun F (Fin n)).invertibleToMatrix b
+    exact isUnit_det_of_invertible _
+  left_inv := by
+    intro
+    ext
+    simp only [transpose_transpose]
+    exact rfl
+  right_inv := by exact congrFun rfl
```
```suggestion
noncomputable def equiv_GL_linearindependent {F : Type*} [Field F] (hn : 0 < n) :
    GL (Fin n) F ≃ { s : Fin n → (Fin n → F) // LinearIndependent F s } where
  toFun M := ⟨transpose M, by
    apply linearIndependent_iff_card_eq_finrank_span.2
    rw [Set.finrank, ← rank_eq_finrank_span_cols, rank_unit]⟩
  invFun M := by
    apply GeneralLinearGroup.mk'' (transpose (M.1))
    rw [show M.1ᵀ = Basis.toMatrix (Pi.basisFun F (Fin n)) (transpose (M.1)ᵀ) by rfl,
      transpose_transpose]
    have : Nonempty (Fin n) := Fin.pos_iff_nonempty.1 hn
    have hdim : Fintype.card (Fin n) = FiniteDimensional.finrank F (Fin n → F) := by
      simp only [Fintype.card_fin, FiniteDimensional.finrank_fintype_fun_eq_card]
    let b := basisOfLinearIndependentOfCardEqFinrank M.2 hdim
    rw [show M = ⇑b by simp only [b, coe_basisOfLinearIndependentOfCardEqFinrank]]
    have : Invertible ((Pi.basisFun F (Fin n)).toMatrix ⇑b) :=
      (Pi.basisFun F (Fin n)).invertibleToMatrix b
    exact isUnit_det_of_invertible _
  left_inv := fun x ↦ Units.ext (ext fun i j ↦ rfl)
  right_inv := by exact congrFun rfl
```

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+noncomputable instance fintype : Fintype (GL (Fin n) 𝔽) := by
+    exact Fintype.ofFinite (GL (Fin n) 𝔽)
+
+theorem card_GL : Fintype.card (GL (Fin n) 𝔽) =
+        ∏ i : (Fin n), (q ^ (n) - q ^ ( i : ℕ )) := by
+    by_cases hn : n = 0
+    · rw [hn]
+      simp only [Fintype.card_unique, Finset.univ_eq_empty, mul_zero, pow_zero,
+      Finset.prod_empty]
+    · rw [Fintype.card_congr (equiv_GL_linearindependent n (Nat.pos_of_ne_zero hn))]
+      exact card_LinearInependent_subtype _ (Nat.le_refl n)
```
```suggestion
theorem card_GL : Fintype.card (GL (Fin n) 𝔽) = ∏ i : (Fin n), (q ^ (n) - q ^ ( i : ℕ )) := by
  by_cases hn : n = 0
  · rw [hn]
    simp only [hn, Fintype.card_unique, Finset.univ_eq_empty, mul_zero, pow_zero,
    Finset.prod_empty]
  · rw [Fintype.card_congr (equiv_GL_linearindependent n (Nat.pos_of_ne_zero hn))]
    exact card_LinearInependent_subtype _ (Nat.le_refl n)
```

ThomasLanard replied:
Here, I removed the `hn` in the `simp only` in the suggested change.

CBirkbeck commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,98 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+
+noncomputable instance fintype : Fintype (GL (Fin n) 𝔽) := by
+    exact Fintype.ofFinite (GL (Fin n) 𝔽)
+
```
Again the bot is complaining that the indentation isnt correct. I've fixed it in the suggestion.

CBirkbeck reviewed (commented):
I've left some comments on the code, it looks pretty good to me, only some minor fixes/golfing.

ThomasLanard commented:
Thank you! I've made the corrections that were suggested.

grunweg commented:
Coming here for issue triage: it seems you're looking for review on this PR; I have labelled it `awaiting-review` accordingly.
Also note that this PR cannot be merged while it is in draft mode. If you consider it ready for review, would you like to un-draft it? (Draft PRs also do not show up in the review queue.)

riccardobrasca commented:
Can you please mark "resolved" the comments that you took into account?

There is also an error related to imports (this should be very easy to resolve). Thanks!

riccardobrasca commented on Mathlib/LinearAlgebra/CliffordAlgebra/Basic.lean:
```diff
@@ -197,6 +197,7 @@ theorem hom_ext {A : Type*} [Semiring A] [Algebra R A] {f g : CliffordAlgebra Q
 #align clifford_algebra.hom_ext CliffordAlgebra.hom_ext
 
 -- This proof closely follows `TensorAlgebra.induction`
+attribute [-instance] IntermediateField.module' normalClosure.algebra in
```
Can you add a comment explaining why this is needed?

riccardobrasca commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,92 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+
+local notation "q" => Fintype.card 𝔽
+
+noncomputable instance {k : ℕ} :
```
Those are the instance that create the problem in the other file, right? What about making them `lemma` and make them local instances in this section?

riccardobrasca replied:
Anyway you have several `{k : ℕ} {s : { s : Fin k → Fin n → 𝔽 // LinearIndependent 𝔽 s }}`. You can just put those in a `variable` line.

riccardobrasca reviewed (changes requested):
Thinking a little bit more about it, maybe you can use [Nat.card](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Cardinal/Finite.html#Nat.card) and completely avoid various `Fintype` instances. It seems to me the best way of doing stuff. Note that you may need the instances in *proofs* (to make the link between `Nat.card` and `Fintype.card`, that has a better API), but there is no need to create global instances.

ThomasLanard commented:
> Thinking a little bit more about it, maybe you can use [Nat.card](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Cardinal/Finite.html#Nat.card) and completely avoid various `Fintype` instances. It seems to me the best way of doing stuff. Note that you may need the instances in _proofs_ (to make the link between `Nat.card` and `Fintype.card`, that has a better API), but there is no need to create global instances.

I changed the Fintype.card into Nat.card and suppressed all the instances (hence, also the modification in the file about Clifford algebra).

riccardobrasca commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,85 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+
+/-- Equivalence between `GL n F` and `n` vectors of length `n` that are linearly independent. Given
+by sending a matrix to its coloumns. -/
+noncomputable def equiv_GL_linearindependent {F : Type*} [Field F] (hn : 0 < n) :
+    GL (Fin n) F ≃ { s : Fin n → (Fin n → F) // LinearIndependent F s } where
+  toFun M := ⟨transpose M, by
+    apply linearIndependent_iff_card_eq_finrank_span.2
+    rw [Set.finrank, ← rank_eq_finrank_span_cols, rank_unit]⟩
+  invFun M := by
+    apply GeneralLinearGroup.mk'' (transpose (M.1))
+    rw [show M.1ᵀ = Basis.toMatrix (Pi.basisFun F (Fin n)) (transpose (M.1)ᵀ) by rfl,
+      transpose_transpose]
+    have : Nonempty (Fin n) := Fin.pos_iff_nonempty.1 hn
+    have hdim : Fintype.card (Fin n) = FiniteDimensional.finrank F (Fin n → F) := by
+      simp only [Fintype.card_fin, FiniteDimensional.finrank_fintype_fun_eq_card]
+    let b := basisOfLinearIndependentOfCardEqFinrank M.2 hdim
+    rw [show M = ⇑b by simp only [b, coe_basisOfLinearIndependentOfCardEqFinrank]]
+    have : Invertible ((Pi.basisFun F (Fin n)).toMatrix ⇑b) :=
+      (Pi.basisFun F (Fin n)).invertibleToMatrix b
+    exact isUnit_det_of_invertible _
+  left_inv := fun x ↦ Units.ext (ext fun i j ↦ rfl)
+  right_inv := by exact congrFun rfl
```
```suggestion
noncomputable def equiv_GL_linearindependent {F : Type*} [Field F] (hn : 0 < n) :
    GL (Fin n) F ≃ { s : Fin n → (Fin n → F) // LinearIndependent F s } where
  toFun M := ⟨transpose M, by
    apply linearIndependent_iff_card_eq_finrank_span.2
    rw [Set.finrank, ← rank_eq_finrank_span_cols, rank_unit]⟩
  invFun M := GeneralLinearGroup.mk'' (transpose (M.1)) <| by
    have : Nonempty (Fin n) := Fin.pos_iff_nonempty.1 hn
    rw [← Basis.coePiBasisFun.toMatrix_eq_transpose,
      ← coe_basisOfLinearIndependentOfCardEqFinrank M.2]
    let b := basisOfLinearIndependentOfCardEqFinrank M.2 (by simp)
    have := (Pi.basisFun F (Fin n)).invertibleToMatrix b
    exact isUnit_det_of_invertible _
  left_inv := fun x ↦ Units.ext (ext fun i j ↦ rfl)
  right_inv := by exact congrFun rfl
```
In general it is better to avoid `rw [show ... by rfl]`: if there is a lemma you can rewrite just use it, otherwise you can add it. The reason is that if at some point someone change the definitions involved your proof will break and it will difficult to maintain (on the other hand, if you rewrite a lemma it will be that lemma that will broke and it is usually easier to fix).

riccardobrasca replied:
Well, it's even already there!

riccardobrasca commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,85 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+  left_inv := fun x ↦ Units.ext (ext fun i j ↦ rfl)
+  right_inv := by exact congrFun rfl
+
+theorem card_GL : Nat.card (GL (Fin n) 𝔽) = ∏ i : (Fin n), (q ^ (n) - q ^ ( i : ℕ )) := by
+  by_cases hn : n = 0
+  · rw [hn]
+    simp only [Nat.card_eq_fintype_card, Fintype.card_unique, Finset.univ_eq_empty, pow_zero,
+      Finset.prod_empty]
+  · rw [Nat.card_congr (equiv_GL_linearindependent n (Nat.pos_of_ne_zero hn))]
+    exact card_LinearInependent_subtype _ (Nat.le_refl n)
```
```suggestion
theorem card_GL : Nat.card (GL (Fin n) 𝔽) = ∏ i : (Fin n), (q ^ (n) - q ^ ( i : ℕ )) := by
  rcases Nat.eq_zero_or_pos n with rfl | hn
  · simp [Nat.card_eq_fintype_card]
  · rw [Nat.card_congr (equiv_GL_linearindependent n hn)]
    exact card_LinearInependent_subtype _ rfl.le
```
Using `rcases` with `rfl` is a nice trick: the first cases is `n=0`, and writing `rfl` replaces `n` with `0` everywhere.

Can you add a docstring to this theorem and add it in the description at the beginning of the file? Thanks!

riccardobrasca commented on Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup.lean:
```diff
@@ -329,4 +331,85 @@ def planeConformalMatrix {R} [Field R] (a b : R) (hab : a ^ 2 + b ^ 2 ≠ 0) :
+/-- Equivalence between `k + 1` vectors of length `n` and `k` vectors of length `n` along with a
+vector in the complement of their span.
+-/
+def inductiveStepEquiv {k : ℕ} :
+    { s : Fin (k + 1) → Fin n → 𝔽 // LinearIndependent 𝔽 s } ≃
+      Σ s : { s : Fin k → Fin n → 𝔽 // LinearIndependent 𝔽 s },
+        ((Submodule.span 𝔽 (Set.range (s : Fin k → Fin n → 𝔽)))ᶜ : Set (Fin n → 𝔽)) where
+  toFun s := by
+    refine ⟨⟨Fin.tail s.val, (linearIndependent_fin_succ.mp s.property).left⟩,
+      ⟨s.val 0, (linearIndependent_fin_succ.mp s.property).right⟩⟩
+  invFun s := by
+    refine ⟨Fin.cons s.2.val s.1.val, linearIndependent_fin_cons.mpr ⟨s.1.property, s.2.property⟩⟩
+  left_inv _ := by simp only [Fin.cons_self_tail, Subtype.coe_eta]
+  right_inv := fun ⟨_, _⟩ => by simp only [Fin.cons_zero, Subtype.coe_eta, Sigma.mk.inj_iff,
+    Fin.tail_cons, heq_eq_eq, and_self]
```
In general we try to avoid producing "data" (like a function) in tactic mode.

riccardobrasca commented:
Even better, we may have *three* files: `Defs`, `Basic` and `Card`, this is what we usually do nowadays. Maybe a preliminary PR doing the split is the way to go.

ThomasLanard commented:
After thinking about it, I found it a bit weird to split the GeneralLinearGroup.lean file into a folder. For instance, there is also a SpecialLinearGroup.lean file and not a folder. Thus instead I added a Card.lean file into the folder Matrix which could contain different statements about cardinals of subsets of matrices. For instance, I added a simple result about the cardinal of the set of matrices over a finite field.

I moved equiv_linearIndependent to Mathlib.LinearAlgebra.LinearIndependent. But I don’t know where to put card_linearInependent.

riccardobrasca commented:
What I am proposing is to split the existing file `GeneraliLinearGroup` into two files `GeneraliLinearGroup.Defs` and ``GeneraliLinearGroup.Basic`, and then add a file `GeneraliLinearGroup.Card` containing your result. It's better to do the splitting in another PR (this kind of PRs are usually reviewed very quickly since nothing has really changed), and then updating this one. I know it looks a slow process, but mathlib is a huge library, and it is better to waste some time but doing things properly.

I can do it you want.

riccardobrasca commented:
Now the file should go in the `GeneralLinearGroup` folder.

riccardobrasca commented on Mathlib.lean:
```diff
@@ -2836,6 +2836,7 @@ import Mathlib.LinearAlgebra.Matrix.Adjugate
 import Mathlib.LinearAlgebra.Matrix.Basis
 import Mathlib.LinearAlgebra.Matrix.BilinearForm
 import Mathlib.LinearAlgebra.Matrix.Block
+import Mathlib.LinearAlgebra.Matrix.Card
```
It will complain here.

CBirkbeck commented:
Thanks for doing all of this!

riccardobrasca commented:
Interesting that we don't need commutativity for the formula to hold (I know commutativity is secretly there, but we don't need to know it).

ThomasLanard commented:
> Interesting that we don't need commutativity for the formula to hold (I know commutativity is secretly there, but we don't need to know it).

I have imported `Mathlib.RingTheory.LittleWedderburn` to have commutativity!

riccardobrasca commented:
> > Interesting that we don't need commutativity for the formula to hold (I know commutativity is secretly there, but we don't need to know it).
> 
> I have imported `Mathlib.RingTheory.LittleWedderburn` to have commutativity! It is the only reason for this import.

Oh, I didn't notice it. I suggest to not import it and prove it for `Field` then.

ThomasLanard commented:
Thank you!

bors r+
